### PR TITLE
So a JPL representation can be obtained of a transformation.

### PR DESCRIPTION
**Pull request by ParthaEth**
Since multiagentmapping and possibly many other program uses JPL notation it is convenient to have a function that can return the transformation vector in JPL notation.

### CODE REVIEW

**Comment by ParthaEth** [comment]
@schneith As we discussed yesterday. Here is the PR. please have a look.

**Comment by HannesSommer** [comment]
These name are two little self-explaining IMO. Why not use something like WXYZ and XYZW in the names?

In fact, for me it was quite surprising that you refer to memory order (or ordering of the default basis) that way. Can you point me to literature that associates JPL and Hamilton with particular memory order? I'm only aware of literature associating it with different multiplication order which makes IMO much more sense. And I don't see any dependence between multiplication and memory order. And btw. as far as I know minkindr is based on Hamiltonian multiplication order! As this other severe difference between JPL and Hamilton actually exists it seems quite dangerous to me to just use the phrase "notation" which could be both.

**Comment by schneith** [comment]
add to whitelist

**Comment by schneith** [comment]
add to whitelist

**Comment by schneith** [comment]
I agree that we should probably be more careful with the namings to avoid any confusion. According to this document: http://www.iri.upc.edu/people/jsola/JoanSola/objectes/notes/kinematics.pdf (page 14, Table 1)

the JPL and Hamilton convention define an order as:
Hamiltonian: q=(q_w, q_v)
JPL: q=(q_v, q_w)

But just a change in parameter order does not correspond to a complete transformation from Hamiltonian to JPL convention (not sure if this is correct though). So what about renaming the function to just
asVectorXyzw
asVectorWxyz
to leave its interpretation open to the user?

**Comment by HannesSommer** [comment]
@schneith Thanks for the interesting reference!

I don't think it should be taken as authority for this type of questions, though. Consider what is written below on the same page : "In fact, some works with the
quaternion’s real component at the end (e.g., the C++ library Eigen) are still considered
as using the Hamilton notation, as long as the other three aspects are maintained." 

Who is considering here? This is totally opaque to me :)!

Apparently the author did not really consider the possibility that some of the binary choices might just not be covered by a label like "Hamiltonian". For memory order this is kind of obvious: Hamilton died way before the first computer was built :). And I don't think we should consider memory order and a formal order on the scalar and vector part (for which Hamilton might have had a preference) the same thing - it can clearly be chosen independently and the consequences are of totally different nature ;).

**Comment by schneith** [comment]
Ok, that makes sense. Should we define in the header of the file what we understand under Hamilton and JPL notion? And then add the interfaces like getCoeffAsJpl and getCoeffAsHamilton?

**Comment by ParthaEth** [comment]
So finally, suppose I want a quaternion of the form [x, y, z, W], with the following meaning. 
`[0, 0, sin(delta/2), cos(delta/2)] =>rotation around Z axis by +delta` + delta is given by right hand rule. Then from a corresponding   kindr::minimal::QuatTransformationTemplate T_A_B. what do I do? i.e. how do I change or not the position and sign of the coefficients of this vector? Q_A_B = T_A_B.asVector().head<4>(). Also it is true that T_A_C = T_A_B * T_B_C, Then to my understanding T_A_B and T_B_C have to be active rotations. IS this true? 

To summarize if I do `Q_A_B = T_A_B.asVector().head<4>()` will I get a Hamiltonian quaternion with ACTIVE rotation with the memory layout as [W, x, y, z]?

**Comment by HannesSommer** [comment]
@schneith , no I would avoid writing comments about some notation to clarify the meaning of these function. When we use identifiers like you suggested: 

asVectorXyzw
asVectorWxyz

there is immediately no question left about the memory order of the result without any comment. We might want to add that it also returns the translation, though. 
I would always prefer putting the necessary information into the identifiers if possible. Because this is what people must read (anyway)!

@ParthaEth , there is some confusion:
* memory order is totally independent the other (much more difficult) binary choices (like active/passive, mult order ..). So yes, T_A_B.asVector().head<4>() will always have the layout [W, x, y, z] because that is how it is defined in the code : https://github.com/ethz-asl/minkindr/blob/jpl_support/minkindr/include/kindr/minimal/implementation/quat-transformation-inl.h#L140.
* about active and passive there seems to be a major confusion: active means you don't switch Frames so T_A_C should not be interpreted as an active quaternion. Importantly a quaternion isn't active or passive by itself. Active and passive are just two ways to interpret any _formal_ rotation on R^3. And a quaternion is representing a formal rotation just as a 3x3 orthogonal matrix does (without being pre associated with an active or passive interpretation).
* As a general hint: do not construct quaternions by hand (as above). This will always be a pain! Because you will probably fail using the right convention. If you need to specify a rotation as human, then use angle axis or a rotation vector and use your library to convert it to the corresponding quaternion. Then check if the quaternion does transform some simple given test vector as expected. If not invert it (nothing else can be necessary). Does that help?

**Comment by ParthaEth** [comment]
Ok. I think I more or less understand the situation. Shall we close this PR?
